File: crypto/src/main/java/net/corda/v5/crypto/SecureHash.java
package net.corda.v5.crypto;

import net.corda.v5.base.annotations.CordaSerializable;
import net.corda.v5.base.types.ByteArrays;
import net.corda.v5.base.types.OpaqueBytes;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.nio.ByteBuffer;

/**
 * Container for a cryptographically secure hash value.
 * Provides utilities for generating a cryptographic hash using different algorithms (currently only SHA-256 supported).
 */
@CordaSerializable
public interface SecureHash {

    /**
     * Returns hexadecimal representation of the hash value.
     */
    @NotNull
    String toHexString();

    /**
     * Returns the first specified number of hexadecimal digits of the {@link SecureHash} value.
     *
     * @param prefixLen The number of characters in the prefix.
     */
    @NotNull
    String prefixChars(int prefixLen);


    /**
     * Returns the first 6 hexadecimal digits of the {@link SecureHash} value.
     */
    @NotNull
    String prefixChars();

    /**
     * Compares the two given instances of the {@link SecureHash} based on the content.
     */
    boolean equals(@Nullable Object other);

    /**
     * Returns a hash code value for the object.
     */
    int hashCode();

    /**
     * Converts a {@link SecureHash} object to a string representation containing the <code>algorithm</code> and hexadecimal
     * representation of the <code>bytes</code> separated by the colon character.
     */
    @NotNull
    String toString();

    @NotNull
    String getAlgorithm();

    @NotNull
    byte[] getHashBytes();

}

File: crypto/src/main/kotlin/net/corda/v5/crypto/SecureHash.kt
package net.corda.v5.crypto

import net.corda.v5.base.annotations.CordaSerializable

/**
 * Container for a cryptographically secure hash value.
 * Provides utilities for generating a cryptographic hash using different algorithms (currently only SHA-256 supported).
 *
 * @param algorithm Hashing algorithm which was used to generate the hash.
 * @param bytes Hash value.
 *
 * @property bytes Hash value
 */
@CordaSerializable
interface SecureHash {

    /**
     * Returns hexadecimal representation of the hash value.
     */
    fun toHexString()

    /**
     * Returns the first [prefixLen] hexadecimal digits of the [SecureHash] value.
     * @param prefixLen The number of characters in the prefix.
     */
    fun prefixChars(prefixLen: Int)

    /**
     * Compares the two given instances of the [SecureHash] based on the content.
     */
    override fun equals(other: Any?): Boolean

    /**
     * Returns a hash code value for the object.
     */
    override fun hashCode(): Int

    /**
     * Converts a [SecureHash] object to a string representation containing the [algorithm] and hexadecimal
     * representation of the [bytes] separated by the colon character.
     */
    override fun toString(): String
}